JUnit XML output: Include timezone information in XML

wmanley opened the pull request:
JUnit writes local-time without a timezone by default.  This means that
JUnit XML parsers default to local-time.  The portal doesn't know what
timezone the user of stbt-rig wants so these timestamps can be interpreted
incorrectly.  This change includes timezone information in the XML output.

There is a small chance of incompatibility caused by this change.  If you
experience this please get in contact at support@stb-tester.com.

TODO:

- [x] Test against real portal
- [x] Test with real Jenkins

wmanley committed:
JUnit XML output: Include timezone information in XML

JUnit writes local-time without a timezone by default.  This means that
JUnit XML parsers default to local-time.  The portal doesn't know what
timezone the user of stbt-rig wants so these timestamps can be interpreted
incorrectly.  This change includes timezone information in the XML output.

There is a small chance of incompatibilty caused by this change.  If you
experience this please get in contact at support@stb-tester.com.

wmanley commented:
Hmm, CI failed with:

>     AttributeError: module 'isort' has no attribute 'SortImports'

Looks like a version locking issue.

drothlis committed:
requirements.txt: Set pylint/astroid/isort versions to match Ubuntu 18.04

To fix "AttributeError: module 'isort' has no attribute 'SortImports'".

drothlis commented:
Travis CI *is* running: https://travis-ci.org/github/stb-tester/stbt-rig
I don't know why the results aren't showing here.

drothlis commented:
Well, it looks like AppVeyor has caught an incompatibility with new versions of pytest:
https://ci.appveyor.com/project/drothlis/stbt-rig/builds/35072163/job/atimtgkwgppxebqj

```
________________________ ERROR collecting test session ________________________
c:\python36-x64\lib\site-packages\pluggy\hooks.py:286: in __call__
    return self._hookexec(self, self.get_hookimpls(), kwargs)
c:\python36-x64\lib\site-packages\pluggy\manager.py:93: in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
c:\python36-x64\lib\site-packages\pluggy\manager.py:87: in <lambda>
    firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
C:\projects\stbt-rig\stbt_rig.py:1227: in pytest_collect_file
    return StbtCollector(path, parent)
c:\python36-x64\lib\site-packages\_pytest\nodes.py:95: in __call__
    warnings.warn(NODE_USE_FROM_PARENT.format(name=self.__name__), stacklevel=2)
E   pytest.PytestDeprecationWarning: Direct construction of StbtCollector has been deprecated, please use StbtCollector.from_parent.
E   See https://docs.pytest.org/en/stable/deprecations.html#node-construction-changed-to-node-from-parent for more details.
```

I'm going to stop working on this now so back over to you @wmanley.

wmanley committed:
pytest: Fix deprecation warning when running against pytest 5.4

See
https://docs.pytest.org/en/stable/deprecations.html#node-construction-changed-to-node-from-parent

wmanley commented:
Reproduced locally with:

```
$ python3 -m venv new-pytest
$ source new-pytest/bin/activate
$ pip install -r requirements.txt
...
$ pytest --version
pytest 6.0.1
$ pytest *.py
...
```

wmanley commented:
Now there are only pylint issues remaining:

```
************* Module stbt_rig
stbt_rig.py:1237:4: [W0223(abstract-method), StbtCollector] Method 'get_closest_marker' is abstract in class 'Node' but is not overridden
stbt_rig.py:1237:4: [W0223(abstract-method), StbtCollector] Method 'gethookproxy' is abstract in class 'FSCollector' but is not overridden
stbt_rig.py:1237:4: [W0223(abstract-method), StbtCollector] Method 'isinitpath' is abstract in class 'FSCollector' but is not overridden
stbt_rig.py:1257:4: [W0223(abstract-method), StbtRemoteTest] Method 'get_closest_marker' is abstract in class 'Node' but is not overridden
```

This is a curious one. I think the `get_closest_marker` issue is casued by a limitation of pylint caused by a curious pattern in pytest.

Source:

```python
class Node(metaclass=NodeMeta):
    ...
    @overload
    def get_closest_marker(self, name: str) -> Optional[Mark]:
        raise NotImplementedError()

    @overload  # noqa: F811
    def get_closest_marker(self, name: str, default: Mark) -> Mark:  # noqa: F811
        raise NotImplementedError()

    def get_closest_marker(  # noqa: F811
        self, name: str, default: Optional[Mark] = None
    ) -> Optional[Mark]:
        """return the first marker matching the name, from closest (for example function) to farther level (for example
        module level).

        :param default: fallback return value of no marker was found
        :param name: name to filter by
        """
        return next(self.iter_markers(name=name), default)
```

So it's repeated in the function definition.  The last definition will take precedence.  Not sure why they've done this.

wmanley commented:
Regarding `gethookproxy` and `isinitpath`, the same issue is reported here: https://github.com/pytest-dev/pytest/issues/7591 and will be fixed in 6.1: https://github.com/pytest-dev/pytest/pull/7648.  

I guess I'll just have to disable this lint :(

wmanley committed:
Disable irrelevant pylint errors

Errors were:

```
************* Module stbt_rig
stbt_rig.py:1237:4: [W0223(abstract-method), StbtCollector] Method 'get_closest_marker' is abstract in class 'Node' but is not overridden
stbt_rig.py:1237:4: [W0223(abstract-method), StbtCollector] Method 'gethookproxy' is abstract in class 'FSCollector' but is not overridden
stbt_rig.py:1237:4: [W0223(abstract-method), StbtCollector] Method 'isinitpath' is abstract in class 'FSCollector' but is not overridden
stbt_rig.py:1257:4: [W0223(abstract-method), StbtRemoteTest] Method 'get_closest_marker' is abstract in class 'Node' but is not overridden
```

The `get_closest_marker` is a pylint false positive because it's defined
multiple times in `class Node`:

```python
class Node(metaclass=NodeMeta):
    ...
    @overload
    def get_closest_marker(self, name: str) -> Optional[Mark]:
        raise NotImplementedError()

    @overload  # noqa: F811
    def get_closest_marker(self, name: str, default: Mark) -> Mark:  # noqa: F811
        raise NotImplementedError()

    def get_closest_marker(  # noqa: F811
        self, name: str, default: Optional[Mark] = None
    ) -> Optional[Mark]:
        """return the first marker matching the name, from closest (for example function) to farther level (for example
        module level).

        :param default: fallback return value of no marker was found
        :param name: name to filter by
        """
        return next(self.iter_markers(name=name), default)
```

The other two don't need to be inherited because they'll never be called:
https://github.com/pytest-dev/pytest/issues/7591 .  This is fixed in
pytest 6.1: https://github.com/pytest-dev/pytest/pull/7648

wmanley committed:
Update Copyright notice year
